In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import * 

In [ ]:
import numpy as np

In [ ]:
coef = np.loadtxt('lab6_practice_coef.txt', delimiter=',')

In [ ]:
print(coef.shape)
print('Num rows:',coef.shape[0])
print('Num cols:',coef.shape[1])

(17, 11)
Num rows: 17
Num cols: 11


In [ ]:
print(coef)

[[  0.   0.   0.   0.   0.   1.   1.   1.   1.   1.   0.]
 [ 85.  92.  45.  27.  31.   0.   0.   0.   0.   0.   1.]
 [ 92.  54.  22.  20.   7.   0.   0.   0.   0.   0.   1.]
 [ 96.  67.  29.  20.  11.   0.   0.   0.   0.   0.   1.]
 [-91. -57. -33. -23. -12.   0.   0.   0.   0.   0.   1.]
 [-99. -75. -26. -24. -41.   0.   0.   0.   0.   0.   1.]
 [-98. -99. -57. -45. -65.   0.   0.   0.   0.   0.   1.]
 [ -1.   0.   0.   0.   0.   1.   0.   0.   0.   0.   0.]
 [  0.  -1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.   0.  -1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.   0.  -1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.   0.  -1.   0.   0.   0.   0.   1.   0.]
 [  1.   0.   0.   0.   0.   1.   0.   0.   0.   0.   0.]
 [  0.   1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.   0.   1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.   0.   1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.   0.   1.   0.   0.   0.   0.   1.   0.]]


In [ ]:
model_lab6 = ConcreteModel()

In [ ]:
N = coef.shape[1]-1
M = coef.shape[0]-1

In [ ]:
obj_coef = coef[0,:-1]
print(obj_coef.shape)
print(obj_coef)

(10,)
[0. 0. 0. 0. 0. 1. 1. 1. 1. 1.]


In [ ]:
constr_coef = coef[1:,:-1]
print(constr_coef.shape)
print(constr_coef)

(16, 10)
[[ 85.  92.  45.  27.  31.   0.   0.   0.   0.   0.]
 [ 92.  54.  22.  20.   7.   0.   0.   0.   0.   0.]
 [ 96.  67.  29.  20.  11.   0.   0.   0.   0.   0.]
 [-91. -57. -33. -23. -12.   0.   0.   0.   0.   0.]
 [-99. -75. -26. -24. -41.   0.   0.   0.   0.   0.]
 [-98. -99. -57. -45. -65.   0.   0.   0.   0.   0.]
 [ -1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.  -1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.  -1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.  -1.   0.   0.   0.   0.   1.   0.]
 [  0.   0.   0.   0.  -1.   0.   0.   0.   0.   1.]
 [  1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.   1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.   1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.   1.   0.   0.   0.   0.   1.   0.]
 [  0.   0.   0.   0.   1.   0.   0.   0.   0.   1.]]


In [ ]:
constr_rhs = coef[1:,-1]
print(constr_rhs.shape)
print(constr_rhs)

(16,)
[1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [ ]:
col_indices_x = np.arange(int(N/2))
print(col_indices_x)

model_lab6.x = Var(col_indices_x)

[0 1 2 3 4]


In [ ]:
col_indices_u = np.arange(int(N/2))
print(col_indices_u)

model_lab6.u = Var(col_indices_u, domain=NonNegativeReals)

[0 1 2 3 4]


In [ ]:
row_indices = np.arange(M)
print(row_indices)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]


In [ ]:
model_lab6.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model_lab6.constraints.add(summation(constr_coef[i][0:int(N/2)],model_lab6.x) + summation(constr_coef[i][int(N/2):N],model_lab6.u)  >= constr_rhs[i])

In [ ]:
model_lab6.objective = Objective(expr = summation(obj_coef[0:int(N/2)],model_lab6.x) + summation(obj_coef[int(N/2):N],model_lab6.u), sense=minimize)

In [ ]:
model_lab6.pprint()

3 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    u_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

2 Var Declarations
    u : Size=5, Index=u_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeReals
          1 :     0 :  None :  None : False :  True : NonNegativeReals
          2 :     0 :  None :  None : False :  True : NonNegativeReals
          3 :     0 :  None :  None : False :  True : NonNegativeReals
          4 :     0 :  None :  None : False :  True : NonNegativeReals
    x : Siz

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_lab6)
print(result)


Problem: 
- Name: unknown
  Lower bound: 0.5479557323
  Upper bound: 0.5479557323
  Number of objectives: 1
  Number of constraints: 17
  Number of variables: 11
  Number of nonzeros: 5
  Sense: minimize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 21
  Error rc: 0
  Time: 0.024240732192993164
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [ ]:
print("2.ANSWER:")
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print("Yes the solver yields an optimal solution")


2.ANSWER:
Solver status: ok
Solver termination condition: optimal
Yes the solver yields an optimal solution


In [ ]:
print("3.ANSWER:")
# display solution
print('\nObjective = ', model_lab6.objective())

print('\nDecision Variables')
for i in range(5):
  print('x[',i,'] = ', model_lab6.x[i].value)
  print('u[',i,'] = ', model_lab6.u[i].value)
print('\n Constraints')
model_lab6.constraints.display()

3.ANSWER:

Objective =  0.547955725

Decision Variables
x[ 0 ] =  -0.046448575
u[ 0 ] =  0.046448575
x[ 1 ] =  0.18612441
u[ 1 ] =  0.18612441
x[ 2 ] =  -0.17131802
u[ 2 ] =  0.17131802
x[ 3 ] =  0.0
u[ 3 ] =  0.0
x[ 4 ] =  -0.14406472
u[ 4 ] =  0.14406472

 Constraints
constraints : Size=16
    Key : Lower : Body               : Upper
      1 :   1.0 : 0.9999996250000001 :  None
      2 :   1.0 : 0.9999997599999997 :  None
      3 :   1.0 :  1.458337769999999 :  None
      4 :   1.0 : 1.0000002550000002 :  None
      5 :   1.0 :  1.000000215000001 :  None
      6 :   1.0 : 5.2549776999999995 :  None
      7 :   0.0 :         0.09289715 :  None
      8 :   0.0 :                0.0 :  None
      9 :   0.0 :         0.34263604 :  None
     10 :   0.0 :                0.0 :  None
     11 :   0.0 :         0.28812944 :  None
     12 :   0.0 :                0.0 :  None
     13 :   0.0 :         0.37224882 :  None
     14 :   0.0 :                0.0 :  None
     15 :   0.0 :               

$\textbf{Approach 2:}$

In this approach we will use some fundamental properties of real numbers. Indeed, for any real number $x$, we can write $x=a-b$ where $a\geq 0, b\geq 0$. 

Hence we can substitute $|x| = |a-b|, \ a\geq 0, b\geq 0$. 

Now we will consider another interesting property of real numbers: 
$|u-v| = u+v \iff u\geq 0, v\geq 0, uv = 0$. 

Using this property we can write $|x| = a+b, \ a\geq 0, \ b\geq 0$. Note that this replacement will imply that $ab=0$. 

Thus we can transform the optimization problem $\textbf{(OP)}$ into a new optimization problem $\textbf{(OP2)}$ where $x_i = a_i - b_i, \text{ and } |x_i| = a_i + b_i, \ \forall i \in \{1,\ldots,5\}$ where $a_i \geq 0, b_i \geq 0, \forall i \in \{1,\ldots,5\}$. Thus we have the new optimization problem $\textbf{(OP2)}$: 

\begin{align}
\min \ (a_1 + b_1) +  (a_2 + b_2) + (a_3 + b_3) + (a_4 + b_4) + (a_5 + b_5) &  \nonumber \\
{\rm{s.t.}} \ 85 (a_1-b_1) + 92 (a_2-b_2) + 45 (a_3-b_3) + 27 (a_4-b_4) + 31 (a_5-b_5) & \geq 1 \nonumber \\
92 (a_1-b_1) + 54 (a_2-b_2) + 22 (a_3-b_3) + 20 (a_4-b_4) + 7 (a_5-b_5) & \geq 1 \nonumber \\
96 (a_1-b_1) + 67 (a_2-b_2) + 29 (a_3-b_3) + 20 (a_4-b_4) + 11 (a_5-b_5) & \geq 1 \nonumber \\
-91 (a_1-b_1) - 57 (a_2-b_2) - 33 (a_3-b_3) - 23 (a_4-b_4) - 12 (a_5-b_5) & \geq 1 \nonumber \\
-99 (a_1-b_1) - 75 (a_2-b_2) - 26 (a_3-b_3) - 24 (a_4-b_4) - 41 (a_5-b_5) & \geq 1 \nonumber \\
-98 (a_1-b_1) - 99 (a_2-b_2) - 57 (a_3-b_3) - 45 (a_4-b_4) - 65 (a_5-b_5) & \geq 1 \nonumber \\ 
a_i \geq 0, \ b_i \geq 0, & \ \forall i \in \{1,\ldots,5\}. \nonumber 
\end{align}


In [ ]:
new_coef = np.loadtxt('lab6_practice_coef_OP2.txt', delimiter=',')

In [ ]:
print(new_coef.shape)
print('Num rows:',new_coef.shape[0])
print('Num cols:',new_coef.shape[1])

(7, 11)
Num rows: 7
Num cols: 11


In [ ]:
print(new_coef)

[[  1.   1.   1.   1.   1.   1.   1.   1.   1.   1.   0.]
 [ 85.  92.  45.  27.  31. -85. -92. -45. -27. -31.   1.]
 [ 92.  54.  22.  20.   7. -92. -54. -22. -20.  -7.   1.]
 [ 96.  67.  29.  20.  11. -96. -67. -29. -20. -11.   1.]
 [-91. -57. -33. -23. -12.  91.  57.  33.  23.  12.   1.]
 [-99. -75. -26. -24. -41.  99.  75.  26.  24.  41.   1.]
 [-98. -99. -57. -45. -65.  98.  99.  57.  45.  65.   1.]]


In [ ]:
new_model_lab6 = ConcreteModel()

In [ ]:
N = new_coef.shape[1]-1
M = new_coef.shape[0]-1
print(N,M)

10 6


In [ ]:
new_obj_coef = new_coef[0,:-1]
print(new_obj_coef.shape)
print(new_obj_coef)

(10,)
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


In [ ]:
new_constr_coef = new_coef[1:,:-1]
print(new_constr_coef.shape)
print(new_constr_coef)

(6, 10)
[[ 85.  92.  45.  27.  31. -85. -92. -45. -27. -31.]
 [ 92.  54.  22.  20.   7. -92. -54. -22. -20.  -7.]
 [ 96.  67.  29.  20.  11. -96. -67. -29. -20. -11.]
 [-91. -57. -33. -23. -12.  91.  57.  33.  23.  12.]
 [-99. -75. -26. -24. -41.  99.  75.  26.  24.  41.]
 [-98. -99. -57. -45. -65.  98.  99.  57.  45.  65.]]


In [ ]:
new_constr_rhs = new_coef[1:,-1]
print(new_constr_rhs.shape)
print(new_constr_rhs)

(6,)
[1. 1. 1. 1. 1. 1.]


In [ ]:
col_indices_a = np.arange(int(N/2))
print(col_indices_a)

new_model_lab6.a = Var(col_indices_a,domain=NonNegativeReals)

[0 1 2 3 4]


In [ ]:
col_indices_b = np.arange(int(N/2))
print(col_indices_b)

new_model_lab6.b = Var(col_indices_b, domain=NonNegativeReals)

[0 1 2 3 4]


In [ ]:
row_indices = np.arange(M)
print(row_indices)
  

[0 1 2 3 4 5]


In [ ]:
new_model_lab6.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  new_model_lab6.constraints.add(summation(new_constr_coef[i][0:int(N/2)],new_model_lab6.a) + summation(new_constr_coef[i][int(N/2):N],new_model_lab6.b)  >= new_constr_rhs[i])

In [ ]:
new_model_lab6.objective = Objective(expr = summation(new_obj_coef[0:int(N/2)],new_model_lab6.a) + summation(new_obj_coef[int(N/2):N],new_model_lab6.b), sense=minimize)

In [ ]:
new_model_lab6.pprint()

3 Set Declarations
    a_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}
    b_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}

2 Var Declarations
    a : Size=5, Index=a_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeReals
          1 :     0 :  None :  None : False :  True : NonNegativeReals
          2 :     0 :  None :  None : False :  True : NonNegativeReals
          3 :     0 :  None :  None : False :  True : NonNegativeReals
          4 :     0 :  None :  None : False :  True : NonNegativeReals
    b : Size=5, Index=b_index
        Key : Lowe

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
new_result = opt_cbc.solve(new_model_lab6)
print(new_result)



Problem: 
- Name: unknown
  Lower bound: 0.5479557323
  Upper bound: 0.5479557323
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 11
  Number of nonzeros: 10
  Sense: minimize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 5
  Error rc: 0
  Time: 0.01837015151977539
Solution: 
- number of solutions: 0
  number of solutions displayed: 0



In [ ]:
print('Solver status:', new_result.solver.status)
print('Solver termination condition:',new_result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', new_model_lab6.objective())

print('\nDecision Variables')
for i in range(5):
  print('a[',i,'] = ', new_model_lab6.a[i].value)
  print('b[',i,'] = ', new_model_lab6.b[i].value)
print('\n Constraints')
new_model_lab6.constraints.display()


Objective =  0.547955725

Decision Variables
a[ 0 ] =  0.0
b[ 0 ] =  0.046448575
a[ 1 ] =  0.18612441
b[ 1 ] =  0.0
a[ 2 ] =  0.0
b[ 2 ] =  0.17131802
a[ 3 ] =  0.0
b[ 3 ] =  0.0
a[ 4 ] =  0.0
b[ 4 ] =  0.14406472

 Constraints
constraints : Size=6
    Key : Lower : Body               : Upper
      1 :   1.0 :  0.999999625000001 :  None
      2 :   1.0 : 0.9999997599999997 :  None
      3 :   1.0 :         1.45833777 :  None
      4 :   1.0 : 1.0000002549999998 :  None
      5 :   1.0 :        1.000000215 :  None
      6 :   1.0 :  5.254977699999998 :  None


**9.ANSWER: Yes the solver yield an optimal solution.**\
**Solver status: ok  \
Solver termination condition: optimal**

**10.ANSWER:  \
Objective =  0.547955725**

Decision Variables \
a[ 0 ] =  0.0 ,
b[ 0 ] =  0.046448575 ,
a[ 1 ] =  0.18612441 ,
b[ 1 ] =  0.0 ,
a[ 2 ] =  0.0 ,
b[ 2 ] =  0.17131802 ,
a[ 3 ] =  0.0 ,
b[ 3 ] =  0.0 ,
a[ 4 ] =  0.0 ,
b[ 4 ] =  0.14406472

 Constraints
constraints : Size=6
    Key : Lower : Body               : Upper
      1 :   1.0 :  0.999999625000001 :  None
      2 :   1.0 : 0.9999997599999997 :  None
      3 :   1.0 :         1.45833777 :  None
      4 :   1.0 : 1.0000002549999998 :  None
      5 :   1.0 :        1.000000215 :  None
      6 :   1.0 :  5.254977699999998 :  None 

In [ ]:
print("11.ANSWER:")
for i in range(5):
  print('x[',i,'] = ','a[',i,']',' - ','b[',i,'] = ', new_model_lab6.a[i].value - new_model_lab6.b[i].value)
  

11.ANSWER:
x[ 0 ] =  a[ 0 ]  -  b[ 0 ] =  -0.046448575
x[ 1 ] =  a[ 1 ]  -  b[ 1 ] =  0.18612441
x[ 2 ] =  a[ 2 ]  -  b[ 2 ] =  -0.17131802
x[ 3 ] =  a[ 3 ]  -  b[ 3 ] =  0.0
x[ 4 ] =  a[ 4 ]  -  b[ 4 ] =  -0.14406472


**12.ANSWER:** \
*Coment>>>* 
After comparing the values of $x_{i} $ , obtained in solving **OPT1** and **OPT2** we found that the values are same in both  cases.
